lib/shelley/src/Cardano/Wallet/Shelley/Pools.hs

```diff
@@ -199,6 +224,41 @@ newStakePoolLayer nl db@DBLayer {..} =
     _knownPools =
         Set.fromList <$> liftIO (atomically listRegisteredPools)
 
+    -- In order to apply the settings, we have to restart the
+    -- metadata sync thread as well to avoid race conditions.
+    _putSettings :: TVar (Async ()) -> Settings -> IO ()
+    _putSettings tvAsyncHandle settings = do
+        killSyncThread
+        gcMetadata
+        writeSettings
+        restartSyncThread
```

Comment:
If anything goes wrong with `gcMetadata` or `writeSettings` we will completely loose the synchronization thread. Also we are currently removing pools from the database and putting the settings in two separate atomic operations. So there's also room for things going wrong here. Suggestion to accommodate these issues:

a) Have `readSettings`, `putSettings` and `delPoolMetadata` in the _same_ database "transaction" (within a single atomic). Otherwise, we open up hard-to-debug concurrency issues, 

b) Move `restartSyncThread` in a `finally` clause so that we can be sure that the thread is restarted no matter what. Could be possibly done with a bracket-style combinator: `bracket killThread restartThread modifySettings`

Reply:
Good catch